src/lib/yang/data.lua

```diff
@@ -0,0 +1,338 @@
+-- Use of this source code is governed by the Apache 2.0 license; see
+-- COPYING.
+module(..., package.seeall)
+
+local ffi = require("ffi")
+local parser = require("lib.yang.parser")
+local schema = require("lib.yang.schema2")
+
+function data_grammar_from_schema(schema)
+   local handlers = {}
+   local function visit(node)
+      local handler = handlers[node.kind]
+      if handler then return handler(node) end
+      return {}
+   end
+   local function visit_body(node)
+      local ret = {}
+      for id,node in pairs(node.body) do
+         for keyword,node in pairs(visit(node)) do
+            assert(not ret[keyword], 'duplicate identifier: '..keyword)
+            ret[keyword] = node
+         end
+      end
+      return ret
+   end
+   function handlers.container(node)
+      if not node.presence then return visit_body(node) end
+      return {[node.id]={type='struct', members=visit_body(node)}}
+   end
+   handlers['leaf-list'] = function(node)
+      return {[node.id]={type='array', element_type=node.type}}
+   end
+   function handlers.list(node)
+      local keys = {}
+      for key in node.key:split(' +') do table.insert(keys, key) end
+      return {[node.id]={type='table', keys=keys, members=visit_body(node)}}
+   end
+   function handlers.leaf(node)
+      return {[node.id]={type='scalar', argument_type=node.type,
+                         default=node.default, mandatory=node.mandatory}}
+   end
+   return {type="struct", members=visit_body(schema)}
+end
+
+ffi.cdef([[
+unsigned long long strtoull (const char *nptr, const char **endptr, int base);
+]])
+
+local function integer_type(min, max)
+   return function(str, k)
+      local str = assert(str, 'missing value for '..k)
+      local start = 1
+      local is_negative
+      local base = 10
+      if str:match('^-') then start, is_negative = 2, true
+      elseif str:match('^+') then start = 2 end
+      if str:match('^0x', start) then base, start = 16, start + 2
+      elseif str:match('^0', start) then base = 8 end
```

Comment:
yes we are required to support it in the sense that the syntax for yang integers in the XML mapping, which we are using as a model for our syntactic mapping, supports octal.